Web Inspector: Add precise editing controls for variation axes in Fonts sidebar panel

rcaliman-apple opened the pull request:
#### d0dcfc4c0662f638e82d05512843ff0000a66c05
<pre>
Web Inspector: Add precise editing controls for variation axes in Fonts sidebar panel
<a href="https://bugs.webkit.org/show_bug.cgi?id=253199">https://bugs.webkit.org/show_bug.cgi?id=253199</a>

Reviewed by Patrick Angle.

Adds input fields to the Fonts siderbar panel to allow editing variation axis values.

HTML number input fields accept and display any character, so we restrict keyboard
input to just the ones that make up number values supported as axis values.

The warning for an invalid input is delayed while a user is typing to allow them to finish.

An input can accept and display and invalid value. It will be marked with a warning.
However, the value written to the stylesheet will be either:
- the default axis value when the input value is invalid;
- the clamped min/max value when the input value is out of bounds;

When focus is moved away from an invalid input,
it will update to reflect the actual value used.

Holding the ArrowUp/ArrowDown keys increments/decrements the value within
the bounds of the min/max range with the corresponding step for that axis.

Holding the Shift key will increase the step by an order of magnitude.
Holding the Alt/Option key will decrease the step by an order of magnitude.

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Models/FontStyles.js:
Ensure values written to CSS properties are strings.

(WI.FontStyles.axisValueToFontPropertyValue):
* Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
(WI.FontDetailsPanel.prototype.update):
* Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.css:
(.details-section &gt; .content &gt; .group &gt; .row.font-variation &gt; input[type=&quot;number&quot;]):
(.details-section &gt; .content &gt; .group &gt; .row.font-variation:has(.warning) &gt; input[type=&quot;number&quot;]):
(.details-section &gt; .content &gt; .group &gt; .row.font-variation &gt; input[type=&quot;number&quot;]::-webkit-inner-spin-button):
(.details-section &gt; .content &gt; .group &gt; .row.font-variation &gt; .value): Deleted.
(.details-section &gt; .content &gt; .group &gt; .row.font-variation:has(.warning) &gt; .value): Deleted.
* Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.js:
(WI.FontVariationDetailsSectionRow):
(WI.FontVariationDetailsSectionRow.prototype.set value):
(WI.FontVariationDetailsSectionRow.prototype._getAxisResolution):
(WI.FontVariationDetailsSectionRow.prototype._showValidity):
(WI.FontVariationDetailsSectionRow.prototype._handleValueBlurEvent):
(WI.FontVariationDetailsSectionRow.prototype._handleValueInputEvent):
(WI.FontVariationDetailsSectionRow.prototype._handleValueKeydownEvent):

Canonical link: <a href="https://commits.webkit.org/261162@main">https://commits.webkit.org/261162@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/595218e4bbe73ebf2be1508d1e81b2deb6f6b6d5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110772 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19855 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/43323 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/2118 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/119589 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21260 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10964 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/1823 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116519 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/15841 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/43323 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/103127 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/97801 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/43323 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/103127 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/12454 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/43323 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/103127 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/13022 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/8983 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18403 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/43323 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7732 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/14973 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/FontStyles.js:
```diff
@@ -80,7 +80,7 @@ WI.FontStyles = class FontStyles
             return value >= 1 ? "italic" : "normal";
         default:
-            return value;
+            return `${value}`;
```
Style: i think we'd prefer `String(value)`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.css:
```diff
@@ -42,16 +42,25 @@
     grid-column: axis-value;
-    text-align: end;
+    height: max-content;
+    width: 6ch;
```
are we using `ch` now?  we've tried to stay pretty consistent with `px` in the past 🤔

is this affected by zooming?  what about if the system's/Safari's default font size is changed?

rcaliman-apple replied:
This was a misguided attempt at ensuring that at least 6 characters from the value are visible without being cut off.
That doesn't work because the font is not monospace. Will revert to using pixels.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.js:
```diff
@@ -142,6 +165,87 @@ WI.FontVariationDetailsSectionRow = class FontVariationDetailsSectionRow extends
+            this.warningMessage = null;
+    }
+
+    _handleValueBlurEvent(event)
```
NIT: we don't normally add an `*Event` suffix to event handlers as the `_handle*` prefix already implies that (and the `event` parameter)

patrickangle commented on Source/WebInspectorUI/UserInterface/Models/FontStyles.js:
```diff
@@ -80,7 +80,7 @@ WI.FontStyles = class FontStyles
             return value >= 1 ? "italic" : "normal";
         default:
-            return value;
+            return `${value}`;
```
Can you elaborate on why this was necessary? It isn't clear to me from the commit message.

rcaliman-apple replied:
This change ensures that a single value written to a dedicated CSS font property like `font-weight` is a string.

In `FontStyles.writeFontVariation()`:
```js
targetPropertyValue = WI.FontStyles.axisValueToFontPropertyValue(tag, value);
...
cssProperty.rawValue = targetPropertyValue;
```

`CSSProperty.rawValue` expects a string (actually its `CSSCompletions.completeUnbalancedValue(value)` called within the `rawValue` setter).

Before, the values we got from the range input in `FontVariationDetailsSectionRow` were strings only by coincidence: reading the `value` of an `<input>` always returns a string.

This patch introduced the `<input type="number">` and dispatching its `valueAsNumber` on change. It also introduced dispatching `WI.FontVariationDetailsSectionRow.Event.VariationValueChanged` with various number fallback values when the input value is invalid (default value, min/max value). This highlighted the type mismatch.

An axis' `value` is a number throughout the classes used here. We should keep it that way. Only when writing to the stylesheet do we need to covert it to a string. That's what `FontStyles.axisValueToFontPropertyValue()` does.

You'll notice in the diff of this patch that we're now also using the `valueAsNumber` from the range input as well.
I will update the test as well.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.js:
```diff
@@ -142,6 +165,87 @@ WI.FontVariationDetailsSectionRow = class FontVariationDetailsSectionRow extends
+        let valueAsNumber = event.target.valueAsNumber;
+
+        this._hasValidValue = false;
+        const validValueFormat = /\-?\d+(.\d+)?/g;
```
NIT: id just inline this

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.js:
```diff
@@ -142,6 +165,87 @@ WI.FontVariationDetailsSectionRow = class FontVariationDetailsSectionRow extends
+            clearTimeout(this._showValidityTimeout);
+
+        // Allow the input to temporarily hold an invalid value without immediately warning the user.
+        this._showValidityTimeout = setTimeout(this._showValidity.bind(this), 750);
```
why `750`?

rcaliman-apple replied:
Removed

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.css:
```diff
@@ -42,16 +42,25 @@
+    height: max-content;
+    width: 6ch;
+    padding-block: 4px;
+    padding-inline: 0.5ch;
```
Does this look odd if you instead just set the padding to be 4px all around? 0.5ch is an odd padding value in my opinion...

rcaliman-apple replied:
Yes, I'll revert to using pixels. `ch` didn't work out as I expected.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.js:
```diff
@@ -142,6 +165,87 @@ WI.FontVariationDetailsSectionRow = class FontVariationDetailsSectionRow extends
+        if (!(this._minimumValue <= valueAsNumber && valueAsNumber <= this._maximumValue)) {
+            this._skipUpdatingInputValue = true;
+
+            this.dispatchEventToListeners(WI.FontVariationDetailsSectionRow.Event.VariationValueChanged, {tag: this._tag, value: Math.min(Math.max(this._minimumValue, valueAsNumber), this._maximumValue)});
```
NIT: `Number.constrain(valueAsNumber, this._minimumValue, this._maximumValue)`

rcaliman-apple replied:
Neat! Used that instead.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.js:
```diff
@@ -142,6 +165,87 @@ WI.FontVariationDetailsSectionRow = class FontVariationDetailsSectionRow extends
+            return;
+        }
+
+        if (!(this._minimumValue <= valueAsNumber && valueAsNumber <= this._maximumValue)) {
```
we dont allow for `valueAsNumber === this._minimumValue` or `valueAsNumber === this._maximumValue`?

NIT: I'd avoid the `(!(` if possible as it's not as easy to read
```
if (valueAsNumber < this._minimumValue || valueAsNumber > this._maximumValue) {
```

rcaliman-apple replied:
>we dont allow for valueAsNumber === this._minimumValue or valueAsNumber === this._maximumValue?

We do. The checks are "lower than or equal".

> NIT: I'd avoid the (!( if possible as it's not as easy to read
`if (valueAsNumber < this._minimumValue || valueAsNumber > this._maximumValue) {`

Agreed, this reads better. Thanks!

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.css:
```diff
@@ -42,16 +42,25 @@
+    width: 6ch;
+    padding-block: 4px;
+    padding-inline: 0.5ch;
+    margin-block-start: -6px;
```
Maybe we need to adjust the gap instead and add padding to the variation axis tag instead? I'm always mildly suspicious of negative margins and padding 😅

rcaliman-apple replied:
I know the feeling :) Adjusting the gap here won't help.

Here's the grid layout for that section without the negative margin in the value input:
<img width="357" alt="Screenshot 2023-03-02 at 5 10 39 PM" src="https://user-images.githubusercontent.com/105809798/222485783-b6c9b5cd-fd0d-42b0-80a9-bfecb90ae5e3.png">

Without the negative margin, the number input is aligned along the block-start with the axis tag and the container for the range input (which is another grid) and its min/max values.

Getting the tag, range input and number input to align to a common baseline is tricky. I might try another grid layout.

The negative margin here is my attempt at ensuring that visual alignment.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.js:
```diff
@@ -142,6 +165,87 @@ WI.FontVariationDetailsSectionRow = class FontVariationDetailsSectionRow extends
+        let valueAsNumber = event.target.valueAsNumber;
+        let step = this._step;
+
+        if (event.shiftKey && event.altKey)
```
i coulda sworn we had a utility for this logic somewhere, but i cant really find it :(

we definitely should tho.  maybe a followup?

rcaliman-apple replied:
There are two other places where the modifier keys change the step:

- [BoxShadowEditor.js#L303-L324](https://github.com/WebKit/WebKit/blob/main/Source/WebInspectorUI/UserInterface/Views/BoxShadowEditor.js#L303-L324)
Operating on an `<input type="text">`
- [SpreadsheetTextField.js#L285-L295](https://github.com/WebKit/WebKit/blob/main/Source/WebInspectorUI/UserInterface/Views/SpreadsheetTextField.js#L285-L295)
Operating on a `<span contenteditable>`

Neither is working on a `<input "type=number">` that natively handles increment/decrement, and which we're trying to use here.

They're not consistent either, one assumes a default step of `1`. For the variation axis, the default step can be `0.1` or `0.01` for axes with very short ranges ([the Skia font](https://v-fonts.com/fonts/skia) is a good example).

As a follow-up we're considering making a custom element that wraps the native HTML number input with the behaviors and constraints we expect in Web Inspector. It might not replace the `<span contenteditable>` case which handles a lot of extra behavior, but it can serve as a predictable number input for other uses, like this one in the Fonts sidebar panel.

dcrousso replied:
i was more just referring to the "look at `event.shiftKey` and multiply by `10`" parts, but i love the idea of making a generalized thing 😃

dcrousso replied:
also btw there's also this logic inside `WI.startEditing`, which i think is used for DOM editing?  im pretty sure i added support for this kinda stuff in the DOM tree too

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.css:
```diff
@@ -42,16 +42,25 @@
 
-.details-section > .content > .group > .row.font-variation:has(.warning) > .value {
+.details-section > .content > .group > .row.font-variation:has(.warning) > input[type="number"] {
     margin-inline-end: 8px; /* Clearance so warning icon doesn't overlap the axis value. */
```
Should we always ensure this clearance? It seems strange that we would cause the slider and text box to shift when a warning is present. This both may be unsettling if the text field you are typing in shifts, as well as will break the visual rhythm as you look up and down a list of these fields since the boxes and sliders may no longer align.

rcaliman-apple replied:
I agree. The clearance does waste a lot of space, though. Applied permanently, it also messes the visual balance on the left and right sides of the Fonts sidebar panel. Applied uniformly, it wastes even more horizontal space.

We need to use a smaller warning icon. 

Devin proposed a more subtle marker, top-aligned and without the yellow background. That can work when the axis has a label which creates the negative space for the icon to occupy. But for axes without labels, just tags, the whole label row is missing so the warning icon will once again overlap with the value input.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.js:
```diff
@@ -142,6 +165,87 @@ WI.FontVariationDetailsSectionRow = class FontVariationDetailsSectionRow extends
+            this.warningMessage = null;
+    }
+
+    _handleValueBlurEvent(event)
```
Maybe `_handleValueTextFieldBlur` to make it clear this is for the value text field and not the value slider? Same for the other handlers below.

rcaliman-apple replied:
Sounds good 👍

I also renamed `_inputRangeElement`  to `_valueSliderElement` for consistency and because 
`_handleInputRangeInput` just sounded funny 😃

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.js:
```diff
@@ -142,6 +165,87 @@ WI.FontVariationDetailsSectionRow = class FontVariationDetailsSectionRow extends
+            clearTimeout(this._showValidityTimeout);
+
+        // Allow the input to temporarily hold an invalid value without immediately warning the user.
+        this._showValidityTimeout = setTimeout(this._showValidity.bind(this), 750);
```
I think this is going to feel like poor performance from a user perspective. I'd personally prefer instant feedback as a type as to the validity of what I've typed, particularly since we start with values in the field.

rcaliman-apple replied:
Ok, I added another check and removed this timeout altogether.

The main case I wanted to avoid is nagging the user while they:
- deleted the contents of the input in expectation of typing a new value
- just started typing a negative value

Now, we skip showing the warning in those cases. We will immediately show the warning for an otherwise invalid value.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.js:
```diff
@@ -142,6 +165,87 @@ WI.FontVariationDetailsSectionRow = class FontVariationDetailsSectionRow extends
+        else if (event.altKey)
+            step = this._step / 10;
+
+        // A number input will only increment and decrement within the range defined by its `min` and `max` values.
```
this is kinda an odd way of handling it IMO

other places that we do things like this instead just do a `Number.constrain(valueAsNumber + step, this._minimumValue, this._maximumValue)`

i.e. i dont think we need to muck with the `this._valueInputElement.step` on every `"keydown"`

rcaliman-apple replied:
I tried it this way and it adds a host of other issues. 

The number input already handles increment/decrement on ArrowUp/ArrowDown with the `step` defined on the element.

We can't set `this._valueInputElement.value` and then `event.preventDefault()` after our own mutation to counteract that because it will prevent triggering the "input" event handler.

If we want to mutate the value, we have to account for the browser's native increment/decrement behavior: subtract or add the input's `step`. With a variable `step` to deal with that's just a lot of needless complexity.

I prefer to give the input element its desired step and leave the browser to do the work.

https://user-images.githubusercontent.com/105809798/222537992-690e735d-27ea-4999-896e-942c5116d15d.mov

patrickangle replied:
I'm less bothered by setting the `step` than Devin is, but shouldn't you just be able to set the element's `step` to `0` to get the field to not do anything special for stepping, or is it "smarter" than that?

dcrousso replied:
yeah that's kinda what i was trying to get at (i.e. dont even use HTML `step` and instead do the work ourselves)

rcaliman-apple replied:
The number input won't allow a step of zero. 😮‍💨  

Frankly, this is too much fighting the native input. If it doesn't do what we need, might as well replace it.
I used a text input instead and explicitly `parseFloat()` on the value to get the number out.

In a follow-up, I'll generalize this mock-number input with the other ones we use.

Aside: someone else's [lament](https://stackoverflow.blog/2022/12/26/why-the-number-input-is-the-worst-input/) on the inadequacy of the number input. Some lessons need to be experienced to be learned. Now I understand why we don't use the builtin number input elsewhere in the code.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.js:
```diff
@@ -142,6 +165,87 @@ WI.FontVariationDetailsSectionRow = class FontVariationDetailsSectionRow extends
+        let valueAsString = event.target.value;
+        let valueAsNumber = event.target.valueAsNumber;
+
+        this._hasValidValue = false;
```
I would either move this to be inside both early-return if statements, or pull the `this._skipUpdatingInputValue` out to here, since you set both later on in the function for the non-early-return cases.

rcaliman-apple replied:
I chose to move it inside the early return statements to clearly communicate when we want that behavior.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.js:
```diff
@@ -142,6 +165,87 @@ WI.FontVariationDetailsSectionRow = class FontVariationDetailsSectionRow extends
+
+            this.dispatchEventToListeners(WI.FontVariationDetailsSectionRow.Event.VariationValueChanged, {tag: this._tag, value: this._defaultValue});
+            return;
+        }
```
Is there really no way we can prevent non-digits from being put in the field? Can we remove the the moment they are typed? Also, the warning message that would currently be used isn't entirely correct for the "this isn't even a number case". Saying the value is out of range undersells the real issue here. If we are going to allow non-digits to exist in this text field, we should have an error that tells you that only digits are allowed, or something to that effect...

patrickangle replied:
Actually, I see in `_handleValueKeydownEvent` we seem to prevent typing non-digits, so is this code ever executed?

rcaliman-apple replied:
Yes, this code still gets executed and catches invalid entries like: `--`, `1-0`, `1-`, `..`, `.1`, `1.`, `1.0.1`, etc.
We _could_  make the checks stricter in `_handleValueKeydownEvent` and prevent accepting characters at the wrong location, but that's just going to turn into whack-a-mole of edge cases to address. Might as well mark them for what they are: invalid values.

Another peculiarity of HTML input number is that they'll accept and display a single dash `-` or dot `.` character, but it won't show up when reading `HTMLInputElement.prototype.value`; it returns an empty string: `""`. The check also catches this empty string value. (The `value` of all `<input>` elements always returns a string. We check that for format, but we use `valueAsNumber` for operations. In the case above, `valueAsNumber` is `NaN`). 

I added an additional check for the empty string value (real or 🤷‍♂️) to avoid showing the warning. This can occur when the user has deleted the contents of the input or just started to type a negative number. We want to avoid nagging the user while they may be in the process of typing.

In response to your other comment, I changed this to immediately show the warning for obvious invalid values.
I think this strikes a good balance between good user experience and accurately representing the state of the input.

PS: on blur of an invalid input caught by this check, its `value` gets replaced with `this._defaultValue`.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.js:
```diff
@@ -142,6 +165,87 @@ WI.FontVariationDetailsSectionRow = class FontVariationDetailsSectionRow extends
+    _showValidity()
+    {
+        if (!this._hasValidValue)
+            this.warningMessage = WI.UIString("Axis value outside of supported range: %s – %s", "A warning that is shown in the Font Details Sidebar when the value for a variation axis is outside of the supported range of values").format(this._formatAxisValueAsString(this._minimumValue), this._formatAxisValueAsString(this._maximumValue));
```
i dont really like the look of how this pushes everything to the left, breaking the gridlike consistency of the section (i.e. the inputs no longer horizontally align).  can we move the ⚠️ icon so that it doesnt shift the inputs?

maybe just add a `top: 0;` and remove the `background-color` and `clip-path` so that it shows up in the topright corner instead?

rcaliman-apple replied:
That can work when the axis has a human-readable label like in the video demo for this patch because the label creates the negative space for the warning icon to show up (provided the label isn't very long).

But there are variable fonts with unlabeled axes. They just have a tag. The whole label row goes away and the warning icon ends up overlapping the number input again.

We could use a smaller icon that fits in the space between the input and the edge of the panel. 
Making the icon smaller will also reduce its visibility. In that case we may want to keep the yellow background.

rcaliman-apple replied:
Here's an attempt at making the icon small enough to fit in the existing gap.
Shown both with and without the yellow background. A warning in the Styles panel for comparison.

@dcrousso @patrickangle what do you think?

## With axis labels
<img width="571" alt="Screenshot 2023-03-02 at 6 14 57 PM" src="https://user-images.githubusercontent.com/105809798/222504678-b026194d-4b5f-46db-a993-177969f47311.png">
<img width="571" alt="Screenshot 2023-03-02 at 6 15 21 PM" src="https://user-images.githubusercontent.com/105809798/222504685-41c07df1-500b-4e81-8bd4-79eab90f4005.png">

## Without axis labels
<img width="571" alt="Screenshot 2023-03-02 at 6 21 07 PM" src="https://user-images.githubusercontent.com/105809798/222504692-2d746362-6f5d-44a8-9c5c-a17d083514ce.png">

<img width="571" alt="Screenshot 2023-03-02 at 6 20 47 PM" src="https://user-images.githubusercontent.com/105809798/222504689-10a79b6a-5e4e-45b2-8962-c45e126e2fb5.png">

rcaliman-apple replied:
We can get away with the same size icon if we drop the additional decoration around it.

<img width="571" alt="Screenshot 2023-03-02 at 6 29 36 PM" src="https://user-images.githubusercontent.com/105809798/222506170-93209fe1-1aa1-4490-9a5a-93ddd575c464.png">
<img width="571" alt="Screenshot 2023-03-02 at 6 29 50 PM" src="https://user-images.githubusercontent.com/105809798/222506183-6b0f1e33-e0b9-4baf-9312-0602f579ee7d.png">

dcrousso replied:
ooooo i like the look of having the entire row be yellow 🤩

patrickangle replied:
I prefer https://user-images.githubusercontent.com/105809798/222506170-93209fe1-1aa1-4490-9a5a-93ddd575c464.png as well with the yellow background remaining and just dropping the explicit shape around the ⚠️ icon.

rcaliman-apple replied:
I dropped the clipped background for the warning icon. Now it fits nicely in the gap. 

Additionally, I removed some of the "magic" negative offsets in favor of using `align-items: center` on the grid row.
The tag label, range input, number input and warning icon now align perfectly 👌

I had to use a hack to extract the min/max labels from the grid flow, otherwise they'd contribute to the slider container's height and mess up the baseline. I intend to fix this this by tweaking the grid definition, but ran out of time today. Will follow-up.

Video of the end result:

https://user-images.githubusercontent.com/105809798/222566993-7dc4ce87-32ad-4fc1-b27a-c5319ec56ac3.mov

rcaliman-apple replied:
>  I intend to fix this this by tweaking the grid definition, but ran out of time today. Will follow-up.

I'll do this in a follow-up: https://bugs.webkit.org/show_bug.cgi?id=253334

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.js:
```diff
@@ -142,6 +165,87 @@ WI.FontVariationDetailsSectionRow = class FontVariationDetailsSectionRow extends
+
+    _handleValueKeydownEvent(event)
+    {
+        // By default, a HTML number input will accept and display any character.
```
"By default" makes me think there is a way to do something other than that default without having to do it ourselves...

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.js:
```diff
@@ -47,30 +54,40 @@ WI.FontVariationDetailsSectionRow = class FontVariationDetailsSectionRow extends
-            this.value = event.target.value;
-            this.dispatchEventToListeners(WI.FontVariationDetailsSectionRow.Event.VariationValueChanged, {tag: event.target.name, value: event.target.value});
+            this.value = event.target.valueAsNumber;
+            this.dispatchEventToListeners(WI.FontVariationDetailsSectionRow.Event.VariationValueChanged, {tag: this._tag, value: this._value});
```
Since we are doing `_handle` methods for the text field, this might be a good time to turn this into its own method as well so that all the input handling code is in the same general place below.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.css:
```diff
@@ -24,13 +24,17 @@
-    grid-template-columns: [axis-tag] var(--axisTagWidth) [slider] 1fr [axis-value] minmax(max-content, var(--axisValueWidth));
-    padding: 5px 20px 5px 12px;
+    column-gap: 6px;
+    row-gap: 0;
```
Are we otherwise inheriting a non-zero row-gap without this?

rcaliman-apple replied:
Not at all. I converted shorthand to longhand mindlessly. Removed.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.js:
```diff
@@ -33,44 +33,57 @@ WI.FontVariationDetailsSectionRow = class FontVariationDetailsSectionRow extends
-            this.value = event.target.value;
-            this.dispatchEventToListeners(WI.FontVariationDetailsSectionRow.Event.VariationValueChanged, {tag: event.target.name, value: event.target.value});
-        }, {signal: abortSignal});
+        this._valueInputElement = this.element.appendChild(document.createElement("input"));
```
I meant to have explicitly suggested this be called `_valueTextFieldElement` to match the suggestion for the handler names in the same spirit of disambiguating between the two things that are both inputs (the text field and the slider)

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Views/DetailsSection.css:
```diff
@@ -252,12 +252,12 @@ body[dir=rtl] .details-section > .header::before {
 }
 
-.details-section > .content > .group > .row:is(.simple, .font-variation):has(.warning) {
+.details-section > .content > .group > .row.simple:has(.warning) {
```
This basically restores `DetailsSection.css` to how it looked like before implementing the recent Fonts sidebar panel input controls.

@dcrousso I know you noted that this is a layering violation and it should live in its own stylesheet, perhaps `DetailsSectionSimpleRow.css`. I'll do that in a separate patch. This one is already sprawling.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.js:
```diff
@@ -142,6 +161,92 @@ WI.FontVariationDetailsSectionRow = class FontVariationDetailsSectionRow extends
+
+        this._hasValidValue = true;
+        this._skipUpdatingInputValue = false;
+        this.warningMessage = null;
```
Is this necessary? It seems like the fact we set `_hasValidValue` to `true` and then call `_showValidity()` below would take care of this for us without having to manage the warning message directly here.

rcaliman-apple replied:
Yes, that's true. Forgot to remove this after the last review.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.js:
```diff
@@ -142,6 +161,92 @@ WI.FontVariationDetailsSectionRow = class FontVariationDetailsSectionRow extends
+
+        if (!/\-?\d+(.\d+)?/.test(valueAsString)) {
+            this._hasValidValue = false;
+            this._skipUpdatingInputValue = true;
```
Am I understanding correctly that these two values will always be in sync, unless the user is not actively focused in the text field? If so perhaps `set value` should just ignore attempts to set the value if the field is currently both invalid and focused?

rcaliman-apple replied:
That's true. We can leverage the focused state of the input and `this._hasValidValue` in the setter. 
Brilliant suggestion! Love this part of code review: when the same behavior can be achieved with less code :) Thanks! 👍

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.css:
```diff
@@ -42,17 +45,17 @@
     grid-column: axis-value;
-    text-align: end;
+    height: max-content;
+    width: 42px;
```
Nit: `var(--axis-value-width)`?

rcaliman-apple committed:
Web Inspector: Add precise editing controls for variation axes in Fonts sidebar panel
https://bugs.webkit.org/show_bug.cgi?id=253199

Reviewed by Patrick Angle.

Adds input fields to the Fonts siderbar panel to allow editing variation axis values.

HTML number input fields accept and display any character, so we restrict keyboard
input to just the ones that make up number values supported as axis values.

The warning for an invalid input is delayed while a user is typing to allow them to finish.

An input can accept and display and invalid value. It will be marked with a warning.
However, the value written to the stylesheet will be either:
- the default axis value when the input value is invalid;
- the clamped min/max value when the input value is out of bounds;

When focus is moved away from an invalid input,
it will update to reflect the actual value used.

Holding the ArrowUp/ArrowDown keys increments/decrements the value within
the bounds of the min/max range with the corresponding step for that axis.

Holding the Shift key will increase the step by an order of magnitude.
Holding the Alt/Option key will decrease the step by an order of magnitude.

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Models/FontStyles.js:
Ensure values written to CSS properties are strings.

(WI.FontStyles.axisValueToFontPropertyValue):
* Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
(WI.FontDetailsPanel.prototype.update):
* Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.css:
(.details-section > .content > .group > .row.font-variation > input[type="number"]):
(.details-section > .content > .group > .row.font-variation:has(.warning) > input[type="number"]):
(.details-section > .content > .group > .row.font-variation > input[type="number"]::-webkit-inner-spin-button):
(.details-section > .content > .group > .row.font-variation > .value): Deleted.
(.details-section > .content > .group > .row.font-variation:has(.warning) > .value): Deleted.
* Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.js:
(WI.FontVariationDetailsSectionRow):
(WI.FontVariationDetailsSectionRow.prototype.set value):
(WI.FontVariationDetailsSectionRow.prototype._getAxisResolution):
(WI.FontVariationDetailsSectionRow.prototype._showValidity):
(WI.FontVariationDetailsSectionRow.prototype._handleValueBlurEvent):
(WI.FontVariationDetailsSectionRow.prototype._handleValueInputEvent):
(WI.FontVariationDetailsSectionRow.prototype._handleValueKeydownEvent):

Canonical link: https://commits.webkit.org/261162@main
